lib/index.js

```diff
@@ -291,104 +290,18 @@ class WebLoginNotSupported extends HttpErrorBase {
-    super(method, target, res, body)
-    this.message = 'Login is not allowed from your IP address'
-    this.code = 'EAUTHIP'
-    Error.captureStackTrace(this, HttpErrorAuthIPAddress)
-  }
-}
-
-class HttpErrorAuthUnknown extends HttpErrorBase {
-  constructor (method, target, res, body) {
-    super(method, target, res, body)
-    this.message = 'Unable to authenticate, need: ' + res.headers.get('www-authenticate')
-    this.code = 'EAUTHUNKNOWN'
-    Error.captureStackTrace(this, HttpErrorAuthUnknown)
-  }
-}
-
-function authHeaders (auth) {
-  const headers = {}
-  if (!auth) return headers
-  if (auth.otp) headers['npm-otp'] = auth.otp
-  if (auth.token) {
-    headers['Authorization'] = 'Bearer ' + auth.token
-  } else if (auth.basic) {
-    const basic = auth.basic.username + ':' + auth.basic.password
-    headers['Authorization'] = 'Basic ' + Buffer.from(basic).toString('base64')
-  }
-  return headers
-}
-
-function fetchJSON (conf) {
-  const fetchOpts = {
-    method: conf.method,
-    headers: Object.assign({}, conf.headers || (conf.auth && authHeaders(conf.auth)) || {})
-  }
-  if (conf.body != null) {
-    fetchOpts.headers['Content-Type'] = 'application/json'
-    fetchOpts.body = JSON.stringify(conf.body)
-  }
-  process.emit('log', 'http', 'request', '→', conf.method || 'GET', conf.target)
-  return fetch.defaults(conf.opts || {})(conf.target, fetchOpts).catch(err => {
-    throw new FetchError(err, conf.method, conf.target)
-  }).then(res => {
-    if (res.headers.has('npm-notice')) {
-      process.emit('warn', 'notice', res.headers.get('npm-notice'))
-    }
-    if (res.headers.get('content-type') === 'application/json') {
-      return res.json().then(content => [res, content])
-    } else {
-      return res.buffer().then(content => {
-        try {
-          return [res, JSON.parse(content)]
-        } catch (_) {
-          return [res, content]
-        }
-      })
-    }
-  }).then(result => {
-    const res = result[0]
-    const content = result[1]
-    const retVal = conf.saveResponse ? result : content
-    process.emit('log', 'http', res.status, `← ${res.statusText} (${conf.target})`)
-    if (res.status === 401 && res.headers.get('www-authenticate')) {
-      const auth = res.headers.get('www-authenticate').split(/,\s*/).map(s => s.toLowerCase())
-      if (auth.indexOf('ipaddress') !== -1) {
-        throw new HttpErrorAuthIPAddress(conf.method, conf.target, res, content)
-      } else if (auth.indexOf('otp') !== -1) {
-        throw new HttpErrorAuthOTP(conf.method, conf.target, res, content)
-      } else {
-        throw new HttpErrorAuthUnknown(conf.method, conf.target, res, content)
-      }
-    } else if (res.status < 200 || res.status >= 300) {
-      throw new HttpErrorGeneral(conf.method, conf.target, res, content)
-    } else {
-      return retVal
-    }
-  })
-}
-
 function packageName (href) {
   try {
     let basePath = url.parse(href).pathname.substr(1)
     if (!basePath.match(/^-/)) {
```

Comment:
This whole if can certainly go away, as all URLs requested in this library start with `-`. This maybe should be replicated over in npm-registry-fetch, however?

This function is used to set the `pkgid` attribute on errors, which is in turn used by npm to provide better error messages with more context to users.